tests/autosave.js

```diff
@@ -0,0 +1,441 @@
+
+			editor.model.change( writer => {
+				writer.addMarker( 'name', { usingOperation: false, affectsData: true, range } );
+			} );
+
+			autosave._flush();
+			sinon.assert.calledOnce( autosave.provider.save );
+
+			editor.model.change( writer => {
+				writer.updateMarker( 'name', { range: range2 } );
+			} );
+
+			autosave._flush();
+
+			sinon.assert.calledTwice( autosave.provider.save );
+		} );
+
+		it( 'should call the save method when some marker affects the data model #3', () => {
+			autosave.provider = {
+				save: sandbox.spy()
+			};
+
+			const range = ModelRange.createIn( editor.model.document.getRoot().getChild( 0 ) );
+
+			editor.model.change( writer => {
+				writer.addMarker( 'marker-not-affecting-data', { usingOperation: false, affectsData: true, range } );
+				writer.addMarker( 'marker-affecting-data', { usingOperation: false, affectsData: false, range } );
+			} );
+
+			autosave._flush();
+			sinon.assert.calledOnce( autosave.provider.save );
+		} );
+
+		it( 'should flush remaining calls after editor\'s destroy', () => {
+			const spy = sandbox.spy();
+			const savedStates = [];
+
+			autosave.provider = {
+				save() {
+					spy();
+
+					savedStates.push( editor.getData() );
+				}
+			};
+
+			editor.model.change( writer => {
+				writer.setSelection( ModelRange.createIn( editor.model.document.getRoot().getChild( 0 ) ) );
+				editor.model.insertContent( new ModelText( 'foo' ), editor.model.document.selection );
+			} );
+
+			editor.model.change( writer => {
+				writer.setSelection( ModelRange.createIn( editor.model.document.getRoot().getChild( 1 ) ) );
+				editor.model.insertContent( new ModelText( 'bar' ), editor.model.document.selection );
+			} );
+
+			return editor.destroy().then( () => {
+				expect( spy.callCount ).to.equal( 2 );
+				expect( savedStates ).to.deep.equal( [
+					'<p>foo</p><p>paragraph2</p>',
+					'<p>foo</p><p>bar</p>',
+				] );
+			} );
+		} );
+
+		it( 'should work after editor\'s destroy with long server\'s action time', () => {
+			sandbox.useFakeTimers();
+			const pendingActions = editor.plugins.get( PendingActions );
+			const serverActionSpy = sinon.spy();
+			const serverActionStub = sinon.stub();
+			serverActionStub.onCall( 0 ).resolves( wait( 500 ).then( serverActionSpy ) );
+
+			autosave.provider = {
+				save: serverActionStub
+			};
+
+			editor.model.change( writer => {
+				writer.setSelection( ModelRange.createIn( editor.model.document.getRoot().getChild( 0 ) ) );
+				editor.model.insertContent( new ModelText( 'foo' ), editor.model.document.selection );
+			} );
+
+			return editor.destroy()
+				.then( () => {
+					expect( pendingActions.isPending ).to.be.true;
+					sandbox.clock.tick( 500 );
+				} )
+				.then( () => {
+					expect( pendingActions.isPending ).to.be.false;
+					sinon.assert.calledOnce( serverActionSpy );
+				} );
+		} );
+	} );
+
+	function wait( time ) {
+		return new Promise( res => {
+			window.setTimeout( res, time );
```

Comment:
Use [sinon fake timer](http://sinonjs.org/releases/v1.17.7/fake-timers/) instead.

Reply:
I use fake timers so that timeout will be called almost immediately.